rclcpp/include/rclcpp/parameter_client.hpp

```diff
@@ -73,28 +75,33 @@ class AsyncParametersClient
     std::function<void(
       std::shared_future<std::vector<rclcpp::parameter::ParameterVariant>>)> callback = nullptr)
   {
-    std::promise<std::vector<rclcpp::parameter::ParameterVariant>> promise_result;
-    auto future_result = promise_result.get_future().share();
+    auto promise_result =
+      std::make_shared<std::promise<std::vector<rclcpp::parameter::ParameterVariant>>>();
+    this->get_parameters_promises_.insert(promise_result);
```

Comment:
I don't think `this` is necessary here since the usage is not in a lambda.

Same for some of the below cases.